onnx/shape_inference/implementation.cc

```diff
@@ -184,7 +184,8 @@ void mergeShapesAndTypes(const TypeProto_Tensor& inferredType, TypeProto_Tensor*
   for (int i = 0; i < inferredType.shape().dim_size(); ++i) {
     const auto& inferredDim = inferredType.shape().dim(i);
     auto* existingDim = existingType->mutable_shape()->mutable_dim(i);
-    if (!existingDim->has_dim_value()) {
+    if ((!existingDim->has_dim_value() && !existingDim->has_dim_param()) ||
+        inferredDim.has_dim_value()) {
```

Comment:
condition
```
(!existingDim->has_dim_value() && !existingDim->has_dim_param()
```
can be folded in for loop on line 179. This way when existing shape is empty we dont have to run through 2 for loops (1 on line 179 and 1 on line 184)

Reply:
Is this what you meant? https://github.com/onnx/onnx/pull/3896/commits/be56060267a7ade6f420152e11c742f8d15a00ab

Note I think we still need `(!existingDim->has_dim_value() && !existingDim->has_dim_param()` here for the case where `existingType` has a known rank (i.e., `existingType->has_shape()` is true) with unknown dimensions.

Reply:
yes thanks for adding line 175. And yes, I agree we still need this check here for the case where existingType has a known rank.